### ISSUE TITLE
Feature: Icon and list Task grouping with preview

### ISSUE THREAD

**Issue by fastrizwaan** [solus-project/budgie-desktop#278]
KDE, cinnamon, windows10 has task grouping with preview

windows 10 Icon task grouping (without numbers)
![0task-grouping](https://cloud.githubusercontent.com/assets/1161682/12060760/74972f6e-af9c-11e5-9ab3-062ec11eb368.png)

cinnamon: http://cinnamon-spices.linuxmint.com/applets/view/16
![task-grouping](https://cloud.githubusercontent.com/assets/1161682/12060733/ed0879b8-af9b-11e5-9991-1f2762b884bd.png)

windows 10 task list grouping with numbers
http://windows.microsoft.com/en-in/windows-vista/group-similar-buttons-on-the-taskbar
![90b3994b-684e-46a3-8dca-e5fdf70bc2f8_14](https://cloud.githubusercontent.com/assets/1161682/12060775/a65fec5c-af9c-11e5-9efb-7d873401d08f.png)

KDE 4 task grouping with preview
![icon-tasks_1](https://cloud.githubusercontent.com/assets/1161682/12060786/ed5eb868-af9c-11e5-8c86-c2ac750b82cf.png)

**Comment by yk11** [comment]
Totally agree (was about to post this myself when I saw it was already posted).

Please ensure that the task groups use the full heigh/width of the panel in case the panel has multiple rows.

**Comment by ikeydoherty** [comment]
Preview would be very nasty to implement, though technically possible. We need to develop a panel->WM protocol to support this, as the panel cannot grab snapshots of the window in a display-server agnostic way.

**Comment by yk11** [comment]
I remember this discussion from a while back.  I think preview is not as important as grouping the apps.

The idea is to have launched pinned apps and launched non-pinned apps show up next to each other, grouped by the app.  This allows for a much cleaner setup.  This is what I use DockX for right now.  Having this handled natively in the panel (and supported properly) makes the setup work.

**Comment by ikeydoherty** [comment]
If someone does a mockup I'm happy to implement it (Windows 10 screenies not counting :P)

**Comment by yk11** [comment]
I can mock it up, but it would be simply listing all the features of DockX and taking screenshots of it.

I actually use DockX without previews, as even the window list looks good.

DockX code is here: https://github.com/M7S/dockbarx (seems to be no longer under active development)

**Comment by ikeydoherty** [comment]
If DockX exists, why not try to make an applet that embeds DockX into budgie-pane? :)

**Comment by yk11** [comment]
XFCE has one.  The issue is that DockX is no longer developed and has some bugs (not enough to stop me from using it, but enough to annoy).

Amazingly, there is NOTHING else that does what it does, and I tried 15+ panel and dock managers.  All the others either force you to have the app already pinned, or lack configuration options, or don't allow for custom launchers, etc.

Taking this functionality and making it a native part of the panel (thus gaining performance, memory savings, etc.) is something that none of the other DEs do.  It is the ONLY think I like about Win7/10, and that saddens me.

**Comment by Grisu-Projects** [comment]
I tried to do a mock-up from stacked icons and how they behave(I am **not** a professional). Hope you like the idea:

### Stacked icons in static state
![Icons in static state](http://i.imgur.com/GwMc16j.png)
### Stacked icons during animation(triggered through hover or click)
![Icons during animation](http://i.imgur.com/T4yhrtN.png)
### Stacked icons full view
![Stacked icons full view](http://i.imgur.com/SjYD5oK.png)

I can do more precise ones later.

**Comment by saligrama** [comment]
I know I'm responding to a 3 month old issue but I for one would really like to see this implemented. It's the one thing keeping me from switching directly to Budgie.

**Comment by ikeydoherty** [comment]
It'll be implemented as and when it's implemented @saligrama - it's currently not a high priority, sorry. Anything currently in TODO.md is. If someone wants to _help_ me implement these things, then that's fine.

**Comment by kq01526** [comment]
> @ikeydoherty wrote:
> Preview would be very nasty to implement, though technically possible. We need to develop a panel->WM protocol to support this, as the panel cannot grab snapshots of the window in a display-server agnostic way.

With Kwin and Compiz it can be done, the following thread has some hints on this:

https://github.com/robint99/mate-dock-applet/issues/49

Does this help?

> @ikeydoherty wrote:
> it's currently not a high priority

Can you please make it a high priority? If you would make it like it's done in Windows 7/8(.1)/10, that would be awesome, because that's what people are used to and like.

Out of the box, no Linux desktop environment currently is offering the the taskbar behavior of MS Windows, even though the Windows taskbar is awesome.

If Budgie would provide a taskbar like the one in MS Windows, it would be huge.

**Comment by bhajneet** [comment]
I also agree that this is one of the major features stopping me from switching from xfce to budgie. In a way, that also stops me from trying solus. I understand that it's not a high priority, and I'm not here to push for that. Instead I want to help flesh out the ideal.

I'll start with a screenshot including dockbarX:
![1485279765](https://cloud.githubusercontent.com/assets/14130567/22259290/b6bad1c4-e232-11e6-8dc5-24b30239c82f.png)

You'll see the menu icon, desktops (1-3), and then the dockbar panel in the bottom left corner.

4 States:
* Open but not active windows with grey border
* Active window with different (green according to gtk theme)
* Loading with another (yellow for me) 
* No border for pinned applications not yet opened. 4 different states to think of.

Splits for 3 States:
* Show up to 3 windows, beyond that it should be obvious you're opening a lot of the same window (This is similar to Windows 7-10 which show a maximum of 3 windows as an indication)
* You can see an example of this for the terminal app

GUI for switching windows:
* No window preview is required, though that would improve things
* No outline of the window being hovered over is required either, though that also would improve experience
* Main requirement is a simple list that lets you activate window or close windows (As shown in screenshot)
* Show active window in one color, mouse hovered windows in another, default background otherwise (Makes sense, but you can decide what is best)

Mouse behavior:
* Single click either launches pinned app or cycles between open apps (Options to cycle through workspaces should defer to budgie settings on whether windows should be shown in all workspaces or not); If there is only one window of an application already open, single click should cycle between restoring and minimizing the window instead.
* Right click could allow users to close all windows, minimize all, tile all (not available in dockbarx), etc.
* Middle click or shift+click can launch another instance of the application

Would love to help, but I'm not a very good programmer. I know that's always annoying, when people have ideas but they cannot help. So I will try to join in the IRC channel from time to time and hopefully pick up some skills. Am also planning to try solus in a VM and tinker around in there as well.

**Comment by ikeydoherty** [comment]
No more features are landing in Budgie 10 I've expressed this a lot.
The only changes will come in Budgie 11.

**Comment by ikeydoherty** [comment]
> Can you please make it a high priority?

No. Budgie has a roadmap and a set of agreed features. I'm not deviating from that. There are far more important things to worry about than providing some kind of compiz preview emulation

**Comment by chifreagle** [comment]
#4

**Comment by fastrizwaan** [comment]
in 2015 dash to panel was not there to show for gnome, but now we have dash to panel and dash to dock extensions which provides preview, could we use code from them?

![screenshot from 2017-09-17 18-09-35](https://user-images.githubusercontent.com/1161682/30520864-98a64a46-9bd3-11e7-8fb3-f96526c68821.png)

https://github.com/jderose9/dash-to-panel

**Comment by ikeydoherty** [comment]
No. They're for GNOME Shell extensions which run inside the compositor process.

**Comment by migueldemoura** [comment]
Is it possible to only show the icons on the tasklist? With the window title I don't have enough panel width for all open programs.

**Comment by Snuggle** [comment]
This would be awesome to see in Budgie 11! Icons can get cluttered pretty quickly and looks pretty clunky if you have many instances of a single application. Even a quick implementation that _doesn't_ show previews would look nicer than current.
![image](https://user-images.githubusercontent.com/26250962/36946645-a763cd8e-1fb7-11e8-8490-e6a2cd19cfec.png)

**Comment by sergioad** [comment]
How is this feature going?

**Comment by mxaddict** [comment]
Is this still being implemented?
It's really annoying to have 3 icons for my browser when I open it...

**Comment by JoshStrobl** [comment]
@mxaddict We have window grouping in the IconTasklist nowadays, you can test it with a fully updated Solus. There's still some issues we need to work, but it's not related to previews (not to come until Budgie 11).

**Comment by mxaddict** [comment]
@JoshStrobl , thanks, gonna compile the master branch (Using Ubuntu Budgie)

**Comment by mxaddict** [comment]
@JoshStrobl , nice I got the new budgie-panel, and looks nice!

Are there any plans to maybe add shortcut + hotkey support (i.e.: Super+1/2/3/4/5/6/7/8/9 to launched pinned apps)

**Comment by JoshStrobl** [comment]
@mxaddict Please, let's not keep necroing this task with random inquiries. Feel free to search for existing feature requests and issues in this tracker.

**Comment by mxaddict** [comment]
@JoshStrobl , ohh sorry, but I assumed it was OK, since this issue is still open, maybe close issues that are resolved?

**Comment by mak3** [comment]
Will Ubuntu Budgie 18.04 LTS get this feature or we need update to 18.10?

**Comment by fossfreedom** [comment]
@mak3 this is an issue tracker for the upstream project. If you are enquiring about project  specific stuff please discuss this on  your project support forum ... in this case for UB it is https://discourse.ubuntubudgie.org
